### DISCUSSION TITLE
Looking for the affected entity in `do_orm_execute` event

### DISCUSSION THREAD

**Discussion by coneybeare** [sqlalchemy/sqlalchemy#11848] [Usage Questions] [answered]
I've got an app which published CRUD events to kafka, using the entity itself and a serializer.

I have a few query patterns, including upserts, and am having troubles getting them to work on all events.

```
"""SQLAlchemy base class for all database models."""

import re
from typing import Any, Dict
from uuid import uuid4

import sqlalchemy.dialects.postgresql as pg
from sqlalchemy import text
from sqlalchemy.orm import declared_attr, Mapped, mapped_column, DeclarativeBase
from sqlalchemy_utils import Timestamp


class Base(DeclarativeBase, Timestamp):
    """SQLAlchemy Base class for all database models."""

    @declared_attr
    def __tablename__(cls) -> str:  # pylint: disable=no-self-argument
        """Define table name for all models as the snake case of the model's name."""
        first_pass = re.sub(r"(.)([A-Z][a-z]+)", r"\1_\2", cls.__name__)  # pylint: disable=no-member
        return re.sub(r"([a-z0-9])([A-Z])", r"\1_\2", first_pass).lower()

    id: Mapped[uuid4] = mapped_column(
        pg.UUID(as_uuid=True), primary_key=True, server_default=text("uuid_generate_v4()")
    )
```

```
"""Define SQLAlchemy model for the FooModel table."""

from typing import Any, Union

from sqlalchemy import String, event, Index
from sqlalchemy.orm import validates, Mapped, mapped_column

from app.models.base import Base


class FooModel(Base):
    """Model for the FooModel table."""

    bar: Mapped[str] = mapped_column(String(36), nullable=False, index=True, unique=True, default=None)

@event.listens_for(FooModel, "after_insert")
def receive_after_insert(_mapper: Any, _connection: Any, target: EventMapping) -> None:
    """Hooks into the after insert event to publish kafka messages."""
    _produce_kafka_message('CREATE', target)


@event.listens_for(EventMapping, "after_update")
def receive_after_update(_mapper: Any, _connection: Any, target: EventMapping) -> None:
    """Hooks into the after update event to publish kafka messages."""
    _produce_kafka_message('UPDATE', target)


@event.listens_for(EventMapping, "after_delete")
def receive_after_delete(_mapper: Any, _connection: Any, target: EventMapping) -> None:
    """Hooks into the after delete event to publish kafka messages."""
    _produce_kafka_message('DELETE', target)
```

And this works for queries like 

```
created_foo = FooModel(bar='abcd')
session.add(created_foo)
self.session.commit()
```

but wen trying to do it with an upsert, and trying to tap into the session `do_orm_execute` hooks, I am having trouble getting the same entity.

```
        values = {bar:'1234'}
        insert_stmt = insert(FooModel).returning(FooModel).values(**values)       
        upsert_stmt = insert_stmt.on_conflict_do_update(
            index_elements=['bar'],
            set_={'bar': values['bar'] + '-dup'}
        )
        created_foo = self.session.scalars(upsert_stmt, execution_options={"populate_existing": True})
        self.session.commit()
```

```
@event.listens_for(Session, "do_orm_execute")
def receive_orm_execute(orm_execute_state: ORMExecuteState):
    """Listen to insert update and delete events"""
    if orm_execute_state.is_insert:
        target = '???' # How can I get the created foo instance?
        _produce_kafka_message('CREATE', target)
        
    if orm_execute_state.is_update:
        target = '???' # How can I get the updated foo instance?
        _produce_kafka_message('UPDATE', target)
    
    if orm_execute_state.is_delete:
          target = '???' # How can I get the deleted foo instance?
        _produce_kafka_message('DELETE', target)
```

I edited this code to be a basic model and setup, hopefully detailed enough to see my question clearly. That question is, how can I get the affected entity/target from the `do_orm_execute` hooks?

**Comment by CaselIT** [comment]
Hi,

As documented in https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.ORMExecuteState you can look for bind_mapper and / or all_mappers

**Reply by coneybeare** [reply]
Isn't that just the mapper, not the instance? the `entity` on it is just the class 

```
(Pdb) orm_execute_state.bind_mapper.entity
<class 'app.models.entities.foo_model.FooModel'>
```

**Reply by CaselIT** [reply]
There isn't any instance if you use insert/update/delete

**Reply by coneybeare** [reply]
ok, if that's true and `do_orm_execute ` is not the right place to listen for the instance on upserts, how else can I get these instances on the events that are not covered by `"after_insert", "after_update", "after_delete"`?

**Reply by CaselIT** [reply]
Sqlalchemy doesn't create instances when running inserts/upserte/delete, since it has no need to.

**Reply by coneybeare** [reply]
I appreciate your responses, but I am sensing some sort of avoidance of helping guide me towards the right direction here.

How would _you_ go about solving the problem of getting the instances as a hook on events, to be able to then serialize and report to a different system, like kafka, or audit records? I was hoping to use events to centralize and DRY up code in the models, rather than sprinkling kafka publishing code all around the code base. Is it just not possible to get a comprehensive solution using events?

**Reply by CaselIT** [reply]
You have access to the values provided to the insert/update etc, so you can choose to send them to your external service, or you could manually try to create the instances to send.

Just note that an update or delete may match many rows so it may be hard to know in advance what instances take part to the statement

**Reply by coneybeare** [reply]
Maybe something like executing, then intercepting the scalars could work?

```
@event.listens_for(Session, "do_orm_execute")
def receive_orm_execute(orm_execute_state: ORMExecuteState) -> Result:
    result = orm_execute_state.invoke_statement()
    if orm_execute_state.is_insert:
        instances = result.scalars().all()
        for instance in instances:
            _produce_kafka_message('CREATE', instance)
    if orm_execute_state.is_update:
        instances = result.scalars().all()
        for instance in instances:
            _produce_kafka_message('UPDATE', instance)
    if orm_execute_state.is_delete:
        instances = result.scalars().all()
        for instance in instances:
            _produce_kafka_message('DELETE', instance)

    return result
```

I haven't gotten this to work yet without affecting the callers, but maybe it would? Not sure if the right direction

**Reply by CaselIT** [reply]
I would use a different interface to send events via your external publishing service, that doesn't take the instances, but the class and the value dict used in the statement. This would allow you to use directly the data provided to the statements

**Reply by zzzeek** [reply, marked as answer]
If you are running server side SQL statements like upserts, this is just a SQL string and some data values you're passing along to your database, there's no concrete "entity" present at all.   You are moving from SQLAlchemy's unit of work methodologies, which deal with in-python entities and have easy event hooks to intercept those entities, to the world of direct SQL execution for things like upserts, where there are by definition no such hooks.

Let's ask why you're using upsert statements.   Because you can create the same data manipulations using regular unit of work - you SELECT the records you want, get a result back, update the ones you have and for the ones that are missing you create new entities and add them to the unit of work.    Why might we not want to do that ?   Two reasons are 1. Atomicity and 2. Performance.    The vast majority of direct SQL upsert uses are for reason 2.  They're faster!  loading objects in with SELECT, taking up all the memory and CPU to build up all those objects, looping through and manipulating them, that's a lot of overhead, with upsert, we skip that whole thing.

So we're foregoing SQLAlchemy's fairly rich but somewhat plodding unit of work paradigm for a more succinct and vastly less client heavy SQL upsert paradigm as a performance optimization.  We could write our whole program without it, but we like the speed and lower client footprint.

When there's performance optimizations in play, we lose things as well.  We're losing the implicit use of real python entities and the ability for there to be event hooks that can watch and intercept all these rich, memory heavy entities moving down the assembly line.

So what happens if we say, oh we also have a Kafka process that *requires those entities to exist or else it can't work*?

Well the most obvious thing is that if it's a requirement that all database operations must also instantiate entities representing each row, the most obvious implication is that we are likely going to be **giving up a large chunk of the optimization we were getting by using upserts** since we're asking to create the entities anyway.

My next thought would be, hey can this Kafka thing maybe be **changed** so that it can work just with something like a class name and a list of particular data fields ?   That way the kind of data we send to an upsert would be useful to our Kafka process as well and we dont have to undo a large chunk of our performance gains by having to create all those objects all over again anyway.

Turns out nope, management shot that down, the Kafka process **must** have a fully instantiated python entity from all affected rows of statements.

So we have here a choice to make.  Do we keep using upserts despite that this means we have to re-instantiate objects from the statement somehow ?

If the answer is yes, then we have to take on that work.  We have to make sure all of our SQL upserts have some way to regenerate entities.   This could be done in one of two ways: 1. We try to recreate entities from what we see in our SQL parameters (very tedious, difficult, and brittle to changes) or we 2. Use returning to get the ORM to generate them for us, as we see at https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#using-returning-with-upsert-statements

We going to use approach 2 because approach 1 would be **extremely** difficult to get working and it wont perform as well as 2 in any case.    

But what does that mean?  It means 1. **all** UPSERT and other DML statements must use RETURNING and 2. the statements result sets must be passed through some function that will send off the returned entities to kafka.

So if this were me, not having done this before, I'd 100% be starting (not necessarily ending, more on that later) with a function in my program that does UPSERTs or whatever else, and **that's the function everyone has to use**.   Let's call it `send_upsert()` for the moment.   The function takes in your statement or maybe just some criteria and data, generates an UPSERT that includes the correct RETURNING, runs the statement, gets the results back , sends the entities off to kafka.     **super boring and explicit**.  But if I'm at work, that's 100% what I'd start with, because it's simple and obvious.    

But now we're worried people will forget to use the function.    Say we have 100 files now that use this function everywhere but new developers are not aware they have to use it, and UPSERTs sneak in that aren't sending fully composed entities to kafka.    Most minimal approach to fix this:   a do_orm_execute() event that looks for a token inside all statements that are observed to be DML upserts (or whatever), and our `send_upsert()` puts the token in the statement.    People who forget to use `send_upsert()` get an immediate runtime error with an informative error message.   **problem solved** - our codebase still has this plodding and quite boring `send_upsert()` thing all throughout but it's **clear and obvious what's going on** and the magic of events does not get involved in any heavy handed way except to poke us when we forget to use it.

So now we have a great function that augments our UPSERTs to use RETURNING, gets objects back, uses kafka.  it's just all over our codebase, and we have a do_orm_execute hook anyway.   Still too boring, we're ready to try for better.     So at this point, **maybe** (and i say maybe because this is in the territory of "should work though I've not tried") we can **embed our send_upsert functionality fully into do_orm_execute**.   But since `send_upsert()` can't do the kafka thing until **after** the SQL statement has executed, we'd need to look into the [do_orm_execute event's re-execution feature](https://docs.sqlalchemy.org/en/20/orm/session_events.html#re-executing-statements).    We'd have an event hook that 1. augments the statement with RETURNING 2. does an internal execution of the statement 3. gets the objects back 4. sends them to kafka  5. does the right thing so that the return value of `session.execute()` on the outside looks like what it should.    This is more complicated and intricate but gets us everything we want.    But why did we have to do something so complicated?   because **we wanted to keep our performance optimization in place but also not lose anything regarding the kafka process**.   It's important here that we not lose sight of the fact that **this is all performance optimization in the first place**.

to sum up, I would:   1. slow down  2. consider do we really need UPSERT, why are we using it 3. decide that our kafka requirement of using full entities absolutely cannot be changed  4. decide to use UPSERT but accept that we need additional scaffolding to keep our requirements in place  5. do a **simple, boring, and more verbose than we'd prefer** version first  6. **iterate** on that approach to see if the more rocket-sciency and esoteric features of do_orm_execute() might be able to automate this.

**Reply by coneybeare** [reply]
Thank you for the detailed response.

The reason we want to use the full entity is that the schemas we've established for the kafka payloads is the same as an api response, which is a full jsonapi spec with relatinoship objects, links, and other things, easily recreated by our serializers with an entity present.

We do have lots of reusable query classes that are single entry points to the DB and can consolidate logic, and we do also have `returning` statements where needed. The upserts however, are a premature optimization looking at it now, as they are being run as async background jobs anyhow, so removing those to a select => update or select => create would allow the object to be in mem, and used for kafka. Our deletes by id could also be refactored to a select -> delete for the same purpose.

So I am thinking that the mapper events ('after_insert', 'after_update', and 'after_delete') can still be used for us after refactoring the query objects to always use the session.add for new and updated records, and session.delete (deletions) methods. And not mess with `do_orm_execute` at all
